question2/odbcdemo.c

```diff
@@ -0,0 +1,288 @@
+#include <sqlext.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <strings.h>
+
+/*
+** Define Some useful defines
+*/
+#if !defined(NULL)
+#define NULL 0
+#endif
+
+/*
+** function: ODBC_error
+**
+** Purpose: Display to stdout current ODBC Errors
+**
+** Arguments: henv _ ODBC Environment handle.
+** hdbc - ODBC Connection Handle error generated on.
+** hstmt - ODBC SQL Handle error generated on.
+**
+** Returns: void
+**
+*/
+void ODBC_error(                 /* Get and print ODBC error messages */
+                SQLHANDLE henv,  /* ODBC Environment */
+                SQLHANDLE hdbc,  /* ODBC Connection Handle */
+                SQLHANDLE hstmt) /* ODBC SQL Handle */
+{
+    UCHAR sqlstate[10];
+    UCHAR errmsg[SQL_MAX_MESSAGE_LENGTH];
+    SQLINTEGER nativeerr;
+    SQLSMALLINT actualmsglen;
+    RETCODE rc;
+loop:
+    rc = SQLError(henv, hdbc, hstmt, (SQLCHAR *)sqlstate, &nativeerr,
+                  (SQLCHAR *)errmsg, SQL_MAX_MESSAGE_LENGTH - 1, &actualmsglen);
+    if (rc == SQL_ERROR) {
+        printf("SQLError failed!\n");
+        return;
+    }
+    if (rc != SQL_NO_DATA_FOUND) {
+        printf("SQLSTATE = %s\n", sqlstate);
+        printf("NATIVE ERROR = %d\n", nativeerr);
+        errmsg[actualmsglen] = '\0';
+        printf("MSG = %s\n\n", errmsg);
+        goto loop;
+    }
+}
+
+/*
+** function: EnvClose
+**
+** Purpose: Frees environment and connection handles.
+**
+** Arguments: henv _ environment handle
+** hdbc - connection to handle
+*/
+void EnvClose(SQLHANDLE henv, SQLHANDLE hdbc) {
+    SQLDisconnect(hdbc);
+    SQLFreeHandle(SQL_HANDLE_DBC, hdbc);
+    SQLFreeHandle(SQL_HANDLE_ENV, henv);
+}
+
+/*
+** function: fgets_wrapper
+**
+** Purpose: Handles newline for fgets from stdin.
+**
+** Arguments: buffer _ character array ptr
+** buflen - max characters
+*/
+char *fgets_wrapper(char *buffer, size_t buflen) {
+    if (fgets(buffer, buflen, stdin) != 0) {
+        buffer[strcspn(buffer, "\n")] = '\0';
+        return buffer;
+    }
+    return 0;
+}
+
+/*
+** Defines used by main program.
+*/
+#define PWD_LEN 32
+#define UID_LEN 32
+#define DSN_LEN 32
+#define USAGE_MSG1 "Usage: %s \n"
+
+/*
+** Program: odbc demo
+**
+** Purpose: ODBC Demo routine.
+*/
+
+typedef struct {
+    UCHAR charCol1[1024];
+    // SQLLEN length1 ;
+    long length1;
+    short length2;
```

Comment:
SQLLEN should be preferred for both of these, as the size of long and short is implementation specific
```suggestion
    SQLLEN length1;
    SQLLEN length2;
```

Reply:
actually, length2 doesn't appear to even be used, so it can probably just be removed
```suggestion
    SQLLEN length1;
```